File: runtime/typescript/frame.ts
/**
 * The RIDL frame envelope: HTTP-free addressing for WebSocket and TCP.
 *
 * This is a browser-safe port of `ridl/framing.py`. It has no Node imports and
 * produces the same canonical UTF-8 JSON bytes as the Rust, Dart, Go, and
 * Python ports.
 */

export const FRAME_VERSION = 1 as const;
export const MAX_FRAME_BYTES = 8 * 1024 * 1024;
export const LENGTH_PREFIX_BYTES = 4;

const FIELD_ORDER = [
  "v",
  "id",
  "t",
  "key",
  "method",
  "path",
  "query",
  "body",
  "code",
  "message",
  "meta",
] as const;
const FRAME_KINDS = new Set(["call", "data", "end", "error", "cancel"] as const);

export type FrameKind = "call" | "data" | "end" | "error" | "cancel";

export class FrameError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "FrameError";
  }
}

export interface Frame {
  readonly v: number;
  readonly id: string;
  readonly t: FrameKind;
  readonly key?: string;
  readonly method?: string;
  readonly path?: string;
  readonly query: ReadonlyArray<readonly [string, string]>;
  readonly body?: unknown;
  readonly hasBody: boolean;
  readonly code?: string;
  readonly message?: string;
  readonly meta: ReadonlyArray<readonly [string, string]>;
}

function bare(t: FrameKind, id: string): Frame {
  return { v: FRAME_VERSION, id, t, query: [], hasBody: false, meta: [] };
}

export function callFrame(
  id: string,
  key: string,
  method: string,
  path: string,
  query: ReadonlyArray<readonly [string, string]> = [],
  body?: { value: unknown },
): Frame {
  return {
    ...bare("call", id),
    key,
    method,
    path,
    query,
    body: body?.value,
    hasBody: body !== undefined,
  };
}

export const dataFrame = (id: string, body: unknown): Frame => ({
  ...bare("data", id),
  body,
  hasBody: true,
});
export const endFrame = (id: string): Frame => bare("end", id);
export const cancelFrame = (id: string): Frame => bare("cancel", id);
export const errorFrame = (
  id: string,
  code: string,
  message?: string,
  body?: { value: unknown },
): Frame => ({
  ...bare("error", id),
  code,
  message,
  body: body?.value,
  hasBody: body !== undefined,
});

export function withMeta(frame: Frame, name: string, value: string): Frame {
  const withoutName = frame.meta.filter(([key]) => key !== name);
  return { ...frame, meta: [...withoutName, [name, value] as const] };
}

export function validate(frame: Frame): void {
  if (frame.v !== FRAME_VERSION) {
    throw new FrameError(`unsupported frame version ${frame.v}`);
  }
  if (!FRAME_KINDS.has(frame.t)) {
    throw new FrameError(`unknown frame type ${String(frame.t)}`);
  }
  if (typeof frame.id !== "string" || !frame.id || [...frame.id].length > 128) {
    throw new FrameError("id must be 1..128 characters");
  }
  if (!Array.isArray(frame.query)) {
    throw new FrameError("query must be an array of [name, value] pairs");
  }
  for (const pair of frame.query) {
    if (
      !Array.isArray(pair) ||
      pair.length !== 2 ||
      typeof pair[0] !== "string" ||
      typeof pair[1] !== "string"
    ) {
      throw new FrameError("each query entry must be a [name, value] pair of strings");
    }
  }

  if (frame.t === "call") {
    if (typeof frame.key !== "string" || !frame.key) {
      throw new FrameError("a call frame needs an operation key");
    }
    if (typeof frame.method !== "string" || !frame.method) {
      throw new FrameError("a call frame needs a method");
    }
    if (typeof frame.path !== "string" || !frame.path.startsWith("/")) {
      throw new FrameError("a call frame needs a path starting with /");
    }
  } else if (frame.key || frame.method || frame.path || frame.query.length) {
    throw new FrameError(`a ${frame.t} frame carries no addressing fields`);
  }

  if (frame.hasBody && frame.body === undefined) {
    throw new FrameError("a present body cannot be JavaScript undefined");
  }
  if (frame.t === "data" && !frame.hasBody) {
    throw new FrameError("a data frame needs a body");
  }
  if (frame.t === "error") {
    if (typeof frame.code !== "string" || !frame.code) {
      throw new FrameError("an error frame needs a code");
    }
    if (frame.message !== undefined && typeof frame.message !== "string") {
      throw new FrameError("an error message must be a string");
    }
  } else if (frame.code !== undefined || frame.message !== undefined) {
    throw new FrameError(`a ${frame.t} frame carries no code or message`);
  }

  if (!Array.isArray(frame.meta)) {
    throw new FrameError("meta must be an array of [name, value] pairs");
  }
  const metaNames = new Set<string>();
  for (const pair of frame.meta) {
    if (
      !Array.isArray(pair) ||
      pair.length !== 2 ||
      typeof pair[0] !== "string" ||
      typeof pair[1] !== "string"
    ) {
      throw new FrameError("each meta entry must be a [name, value] pair of strings");
    }
    if (metaNames.has(pair[0])) {
      throw new FrameError(`duplicate meta member ${pair[0]}`);
    }
    metaNames.add(pair[0]);
  }
}

function toObject(frame: Frame): Record<string, unknown> {
  const raw: Record<string, unknown> = { v: frame.v, id: frame.id, t: frame.t };
  if (frame.t === "call") {
    raw.key = frame.key;
    raw.method = frame.method;
    raw.path = frame.path;
    if (frame.query.length) raw.query = frame.query.map(([key, value]) => [key, value]);
  }
  if (frame.hasBody) raw.body = frame.body;
  if (frame.t === "error") {
    raw.code = frame.code;
    if (frame.message !== undefined) raw.message = frame.message;
  }
  if (frame.meta.length) {
    raw.meta = Object.fromEntries(
      [...frame.meta].sort(([left], [right]) => left.localeCompare(right)),
    );
  }

  const ordered: Record<string, unknown> = {};
  for (const name of FIELD_ORDER) {
    if (name in raw) ordered[name] = raw[name];
  }
  return ordered;
}

export function encode(frame: Frame): Uint8Array {
  validate(frame);
  let text: string;
  try {
    const encoded = JSON.stringify(toObject(frame));
    if (encoded === undefined) {
      throw new FrameError("frame is not JSON-encodable");
    }
    text = encoded;
  } catch (cause) {
    if (cause instanceof FrameError) throw cause;
    throw new FrameError(`frame is not JSON-encodable: ${String(cause)}`);
  }
  const bytes = new TextEncoder().encode(text);
  if (bytes.length > MAX_FRAME_BYTES) {
    throw new FrameError(`frame is ${bytes.length} bytes, over the ${MAX_FRAME_BYTES} limit`);
  }
  return bytes;
}

export function encodeTcp(frame: Frame): Uint8Array {
  const payload = encode(frame);
  const out = new Uint8Array(LENGTH_PREFIX_BYTES + payload.length);
  new DataView(out.buffer).setUint32(0, payload.length, false);
  out.set(payload, LENGTH_PREFIX_BYTES);
  return out;
}

export function decode(payload: Uint8Array | string): Frame {
  let text: string;
  if (typeof payload === "string") {
    const size = new TextEncoder().encode(payload).length;
    if (size > MAX_FRAME_BYTES) {
      throw new FrameError(`frame is ${size} bytes, over the ${MAX_FRAME_BYTES} limit`);
    }
    text = payload;
  } else {
    if (payload.length > MAX_FRAME_BYTES) {
      throw new FrameError(`frame is ${payload.length} bytes, over the ${MAX_FRAME_BYTES} limit`);
    }
    try {
      text = new TextDecoder("utf-8", { fatal: true }).decode(payload);
    } catch (cause) {
      throw new FrameError(`frame is not UTF-8: ${String(cause)}`);
    }
  }

  let raw: unknown;
  try {
    raw = JSON.parse(text);
  } catch (cause) {
    throw new FrameError(`frame is not JSON: ${String(cause)}`);
  }
  if (typeof raw !== "object" || raw === null || Array.isArray(raw)) {
    throw new FrameError("a frame must be a JSON object");
  }
  const obj = raw as Record<string, unknown>;
  const unknown = Object.keys(obj).filter(
    (key) => !(FIELD_ORDER as readonly string[]).includes(key),
  );
  if (unknown.length) {
    throw new FrameError(`unknown frame member(s): ${unknown.sort().join(", ")}`);
  }

  const query: Array<readonly [string, string]> = [];
  if (obj.query !== undefined) {
    if (!Array.isArray(obj.query)) {
      throw new FrameError("query must be an array of [name, value] pairs");
    }
    for (const pair of obj.query) {
      if (
        !Array.isArray(pair) ||
        pair.length !== 2 ||
        pair.some((value) => typeof value !== "string")
      ) {
        throw new FrameError("each query entry must be a [name, value] pair of strings");
      }
      query.push([pair[0] as string, pair[1] as string]);
    }
  }

  const meta: Array<readonly [string, string]> = [];
  if (obj.meta !== undefined) {
    if (typeof obj.meta !== "object" || obj.meta === null || Array.isArray(obj.meta)) {
      throw new FrameError("meta must be an object");
    }
    for (const [key, value] of Object.entries(obj.meta as Record<string, unknown>).sort(
      ([left], [right]) => left.localeCompare(right),
    )) {
      if (typeof value !== "string") {
        throw new FrameError(`meta.${key} must be a string`);
      }
      meta.push([key, value]);
    }
  }

  const frame: Frame = {
    v: typeof obj.v === "number" ? obj.v : 0,
    id: typeof obj.id === "string" ? obj.id : "",
    t: obj.t as FrameKind,
    key: obj.key as string | undefined,
    method: obj.method as string | undefined,
    path: obj.path as string | undefined,
    query,
    body: obj.body,
    hasBody: "body" in obj,
    code: obj.code as string | undefined,
    message: obj.message as string | undefined,
    meta,
  };
  validate(frame);
  return frame;
}

export function decodeStream(buffer: Uint8Array): {
  frames: Frame[];
  rest: Uint8Array;
} {
  const frames: Frame[] = [];
  const view = new DataView(buffer.buffer, buffer.byteOffset, buffer.byteLength);
  let offset = 0;
  while (buffer.length - offset >= LENGTH_PREFIX_BYTES) {
    const length = view.getUint32(offset, false);
    if (length > MAX_FRAME_BYTES) {
      throw new FrameError(
        `declared frame length ${length} is over the ${MAX_FRAME_BYTES} limit`,
      );
    }
    const start = offset + LENGTH_PREFIX_BYTES;
    if (buffer.length - start < length) break;
    frames.push(decode(buffer.subarray(start, start + length)));
    offset = start + length;
  }
  return { frames, rest: buffer.subarray(offset) };
}

export class Correlator {
  #next = 0n;
  readonly #prefix: string;

  constructor(prefix = "") {
    this.#prefix = prefix;
  }

  take(): string {
    this.#next += 1n;
    return this.#prefix ? `${this.#prefix}${this.#next}` : String(this.#next);
  }
}

File: runtime/rust/frame.rs
//! The RIDL frame envelope: HTTP-free addressing for WebSocket and TCP.
//!
//! This is a standard-library-plus-serde port of `ridl/framing.py`. The
//! fixtures under `examples/frames/` are the byte contract shared by Rust,
//! Dart, Go, TypeScript, and Python. Nothing in this module performs I/O.

use std::collections::BTreeSet;

use serde_json::Value;

pub const FRAME_VERSION: u8 = 1;
pub const MAX_FRAME_BYTES: usize = 8 * 1024 * 1024;
pub const LENGTH_PREFIX_BYTES: usize = 4;

const FIELD_ORDER: [&str; 11] = [
    "v", "id", "t", "key", "method", "path", "query", "body", "code", "message", "meta",
];

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum FrameKind {
    Call,
    Data,
    End,
    Error,
    Cancel,
}

impl FrameKind {
    #[must_use]
    pub const fn as_str(self) -> &'static str {
        match self {
            Self::Call => "call",
            Self::Data => "data",
            Self::End => "end",
            Self::Error => "error",
            Self::Cancel => "cancel",
        }
    }

    #[must_use]
    pub fn parse(raw: &str) -> Option<Self> {
        Some(match raw {
            "call" => Self::Call,
            "data" => Self::Data,
            "end" => Self::End,
            "error" => Self::Error,
            "cancel" => Self::Cancel,
            _ => return None,
        })
    }
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct FrameError(pub String);

impl std::fmt::Display for FrameError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        formatter.write_str(&self.0)
    }
}

impl std::error::Error for FrameError {}

fn err<T>(message: impl Into<String>) -> Result<T, FrameError> {
    Err(FrameError(message.into()))
}

/// One message on a framed transport.
///
/// `None` means the `body` member is absent; `Some(Value::Null)` means the
/// member is present with JSON `null`.
#[derive(Clone, Debug, PartialEq)]
pub struct Frame {
    pub v: u8,
    pub id: String,
    pub kind: FrameKind,
    pub key: Option<String>,
    pub method: Option<String>,
    pub path: Option<String>,
    pub query: Vec<(String, String)>,
    pub body: Option<Value>,
    pub code: Option<String>,
    pub message: Option<String>,
    pub meta: Vec<(String, String)>,
}

impl Frame {
    fn bare(kind: FrameKind, id: impl Into<String>) -> Self {
        Self {
            v: FRAME_VERSION,
            id: id.into(),
            kind,
            key: None,
            method: None,
            path: None,
            query: Vec::new(),
            body: None,
            code: None,
            message: None,
            meta: Vec::new(),
        }
    }

    #[must_use]
    pub fn call(
        id: impl Into<String>,
        key: impl Into<String>,
        method: impl Into<String>,
        path: impl Into<String>,
        query: Vec<(String, String)>,
        body: Option<Value>,
    ) -> Self {
        let mut frame = Self::bare(FrameKind::Call, id);
        frame.key = Some(key.into());
        frame.method = Some(method.into());
        frame.path = Some(path.into());
        frame.query = query;
        frame.body = body;
        frame
    }

    #[must_use]
    pub fn data(id: impl Into<String>, body: Value) -> Self {
        let mut frame = Self::bare(FrameKind::Data, id);
        frame.body = Some(body);
        frame
    }

    #[must_use]
    pub fn end(id: impl Into<String>) -> Self {
        Self::bare(FrameKind::End, id)
    }

    #[must_use]
    pub fn cancel(id: impl Into<String>) -> Self {
        Self::bare(FrameKind::Cancel, id)
    }

    #[must_use]
    pub fn error(
        id: impl Into<String>,
        code: impl Into<String>,
        message: Option<String>,
    ) -> Self {
        let mut frame = Self::bare(FrameKind::Error, id);
        frame.code = Some(code.into());
        frame.message = message;
        frame
    }

    /// Set one out-of-band string value. Reusing a name replaces its previous
    /// value so encoding can never create duplicate JSON object members.
    #[must_use]
    pub fn with_meta(mut self, name: impl Into<String>, value: impl Into<String>) -> Self {
        let name = name.into();
        let value = value.into();
        if let Some((_, existing)) = self.meta.iter_mut().find(|(key, _)| key == &name) {
            *existing = value;
        } else {
            self.meta.push((name, value));
        }
        self
    }

    #[must_use]
    pub fn meta_get(&self, name: &str) -> Option<&str> {
        self.meta
            .iter()
            .find(|(key, _)| key == name)
            .map(|(_, value)| value.as_str())
    }

    pub fn validate(&self) -> Result<(), FrameError> {
        if self.v != FRAME_VERSION {
            return err(format!("unsupported frame version {}", self.v));
        }
        if self.id.is_empty() || self.id.chars().count() > 128 {
            return err("id must be 1..128 characters");
        }
        match self.kind {
            FrameKind::Call => {
                if self.key.as_deref().unwrap_or_default().is_empty() {
                    return err("a call frame needs an operation key");
                }
                if self.method.as_deref().unwrap_or_default().is_empty() {
                    return err("a call frame needs a method");
                }
                if !self.path.as_deref().is_some_and(|path| path.starts_with('/')) {
                    return err("a call frame needs a path starting with /");
                }
            }
            _ => {
                if self.key.is_some()
                    || self.method.is_some()
                    || self.path.is_some()
                    || !self.query.is_empty()
                {
                    return err(format!(
                        "a {} frame carries no addressing fields",
                        self.kind.as_str()
                    ));
                }
            }
        }
        if self.kind == FrameKind::Data && self.body.is_none() {
            return err("a data frame needs a body");
        }
        if self.kind == FrameKind::Error {
            if self.code.as_deref().unwrap_or_default().is_empty() {
                return err("an error frame needs a code");
            }
        } else if self.code.is_some() || self.message.is_some() {
            return err(format!(
                "a {} frame carries no code or message",
                self.kind.as_str()
            ));
        }

        let mut names = BTreeSet::new();
        for (name, _) in &self.meta {
            if !names.insert(name.as_str()) {
                return err(format!("duplicate meta member {name}"));
            }
        }
        Ok(())
    }

    fn write_json(&self, output: &mut String) -> Result<(), FrameError> {
        fn push_string(output: &mut String, value: &str) -> Result<(), FrameError> {
            let encoded = serde_json::to_string(value)
                .map_err(|error| FrameError(format!("encode: {error}")))?;
            output.push_str(&encoded);
            Ok(())
        }

        fn push_value(output: &mut String, value: &Value) -> Result<(), FrameError> {
            let encoded = serde_json::to_string(value)
                .map_err(|error| FrameError(format!("encode: {error}")))?;
            output.push_str(&encoded);
            Ok(())
        }

        output.push('{');
        output.push_str("\"v\":");
        output.push_str(&self.v.to_string());
        output.push_str(",\"id\":");
        push_string(output, &self.id)?;
        output.push_str(",\"t\":");
        push_string(output, self.kind.as_str())?;

        if self.kind == FrameKind::Call {
            output.push_str(",\"key\":");
            push_string(output, self.key.as_deref().unwrap_or_default())?;
            output.push_str(",\"method\":");
            push_string(output, self.method.as_deref().unwrap_or_default())?;
            output.push_str(",\"path\":");
            push_string(output, self.path.as_deref().unwrap_or_default())?;
            if !self.query.is_empty() {
                output.push_str(",\"query\":[");
                for (index, (name, value)) in self.query.iter().enumerate() {
                    if index > 0 {
                        output.push(',');
                    }
                    output.push('[');
                    push_string(output, name)?;
                    output.push(',');
                    push_string(output, value)?;
                    output.push(']');
                }
                output.push(']');
            }
        }

        if let Some(body) = &self.body {
            output.push_str(",\"body\":");
            push_value(output, body)?;
        }

        if self.kind == FrameKind::Error {
            output.push_str(",\"code\":");
            push_string(output, self.code.as_deref().unwrap_or_default())?;
            if let Some(message) = &self.message {
                output.push_str(",\"message\":");
                push_string(output, message)?;
            }
        }

        if !self.meta.is_empty() {
            let mut meta: Vec<&(String, String)> = self.meta.iter().collect();
            meta.sort_by(|left, right| left.0.cmp(&right.0));
            output.push_str(",\"meta\":{");
            for (index, (name, value)) in meta.iter().enumerate() {
                if index > 0 {
                    output.push(',');
                }
                push_string(output, name)?;
                output.push(':');
                push_string(output, value)?;
            }
            output.push('}');
        }

        output.push('}');
        Ok(())
    }

    pub fn encode(&self) -> Result<Vec<u8>, FrameError> {
        self.validate()?;
        let mut text = String::new();
        self.write_json(&mut text)?;
        let bytes = text.into_bytes();
        if bytes.len() > MAX_FRAME_BYTES {
            return err(format!(
                "frame is {} bytes, over the {MAX_FRAME_BYTES} limit",
                bytes.len()
            ));
        }
        Ok(bytes)
    }

    pub fn encode_tcp(&self) -> Result<Vec<u8>, FrameError> {
        let payload = self.encode()?;
        let mut output = Vec::with_capacity(LENGTH_PREFIX_BYTES + payload.len());
        output.extend_from_slice(&(payload.len() as u32).to_be_bytes());
        output.extend_from_slice(&payload);
        Ok(output)
    }

    pub fn decode(payload: &[u8]) -> Result<Self, FrameError> {
        if payload.len() > MAX_FRAME_BYTES {
            return err(format!(
                "frame is {} bytes, over the {MAX_FRAME_BYTES} limit",
                payload.len()
            ));
        }
        let value: Value = serde_json::from_slice(payload)
            .map_err(|error| FrameError(format!("frame is not JSON: {error}")))?;
        let Value::Object(object) = value else {
            return err("a frame must be a JSON object");
        };

        let mut unknown: Vec<&str> = object
            .keys()
            .map(String::as_str)
            .filter(|key| !FIELD_ORDER.contains(key))
            .collect();
        unknown.sort_unstable();
        if !unknown.is_empty() {
            return err(format!("unknown frame member(s): {}", unknown.join(", ")));
        }

        let raw_version = object
            .get("v")
            .and_then(Value::as_u64)
            .ok_or_else(|| FrameError("v has the wrong type".into()))?;
        let version = u8::try_from(raw_version)
            .map_err(|_| FrameError(format!("unsupported frame version {raw_version}")))?;
        let kind = required_string(&object, "t")
            .and_then(|value| {
                FrameKind::parse(value)
                    .ok_or_else(|| FrameError(format!("unknown frame type {value}")))
            })?;

        let mut query = Vec::new();
        if let Some(raw) = object.get("query") {
            let pairs = raw
                .as_array()
                .ok_or_else(|| FrameError("query must be an array of [name, value] pairs".into()))?;
            for pair in pairs {
                let values = pair.as_array().ok_or_else(|| {
                    FrameError("each query entry must be a [name, value] pair".into())
                })?;
                if values.len() != 2 {
                    return err("each query entry must be a [name, value] pair");
                }
                let name = values[0].as_str().ok_or_else(|| {
                    FrameError("each query entry must be a pair of strings".into())
                })?;
                let value = values[1].as_str().ok_or_else(|| {
                    FrameError("each query entry must be a pair of strings".into())
                })?;
                query.push((name.to_owned(), value.to_owned()));
            }
        }

        let mut meta = Vec::new();
        if let Some(raw) = object.get("meta") {
            let map = raw
                .as_object()
                .ok_or_else(|| FrameError("meta must be an object".into()))?;
            for (name, value) in map {
                let value = value
                    .as_str()
                    .ok_or_else(|| FrameError(format!("meta.{name} must be a string")))?;
                meta.push((name.clone(), value.to_owned()));
            }
            meta.sort_by(|left, right| left.0.cmp(&right.0));
        }

        let frame = Self {
            v: version,
            id: required_string(&object, "id")?.to_owned(),
            kind,
            key: optional_string(&object, "key")?,
            method: optional_string(&object, "method")?,
            path: optional_string(&object, "path")?,
            query,
            body: object.get("body").cloned(),
            code: optional_string(&object, "code")?,
            message: optional_string(&object, "message")?,
            meta,
        };
        frame.validate()?;
        Ok(frame)
    }
}

fn required_string<'a>(
    object: &'a serde_json::Map<String, Value>,
    name: &str,
) -> Result<&'a str, FrameError> {
    object
        .get(name)
        .and_then(Value::as_str)
        .ok_or_else(|| FrameError(format!("{name} has the wrong type")))
}

fn optional_string(
    object: &serde_json::Map<String, Value>,
    name: &str,
) -> Result<Option<String>, FrameError> {
    match object.get(name) {
        None => Ok(None),
        Some(value) => value
            .as_str()
            .map(|value| Some(value.to_owned()))
            .ok_or_else(|| FrameError(format!("{name} has the wrong type"))),
    }
}

pub fn decode_stream(buffer: &[u8]) -> Result<(Vec<Frame>, usize), FrameError> {
    let mut frames = Vec::new();
    let mut offset = 0;
    while buffer.len() - offset >= LENGTH_PREFIX_BYTES {
        let length = u32::from_be_bytes(
            buffer[offset..offset + LENGTH_PREFIX_BYTES]
                .try_into()
                .expect("slice length checked"),
        ) as usize;
        if length > MAX_FRAME_BYTES {
            return err(format!(
                "declared frame length {length} is over the {MAX_FRAME_BYTES} limit"
            ));
        }
        let start = offset + LENGTH_PREFIX_BYTES;
        if buffer.len() - start < length {
            break;
        }
        frames.push(Frame::decode(&buffer[start..start + length])?);
        offset = start + length;
    }
    Ok((frames, offset))
}

#[derive(Debug, Default)]
pub struct Correlator {
    prefix: String,
    next: u64,
}

impl Correlator {
    #[must_use]
    pub fn new(prefix: impl Into<String>) -> Self {
        Self {
            prefix: prefix.into(),
            next: 0,
        }
    }

    pub fn try_take(&mut self) -> Result<String, FrameError> {
        self.next = self
            .next
            .checked_add(1)
            .ok_or_else(|| FrameError("correlation id counter exhausted".into()))?;
        if self.prefix.is_empty() {
            Ok(self.next.to_string())
        } else {
            Ok(format!("{}{}", self.prefix, self.next))
        }
    }

    #[must_use]
    pub fn take(&mut self) -> String {
        self.try_take()
            .expect("correlation id counter exhausted before process restart")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn call_frame_matches_the_canonical_bytes() {
        let frame = Frame::call(
            "1",
            "walk_matter",
            "POST",
            "/v1/matters/abc/walk",
            vec![("include".into(), "1".into())],
            Some(json!({"choice_id": "c"})),
        );
        assert_eq!(
            String::from_utf8(frame.encode().unwrap()).unwrap(),
            r#"{"v":1,"id":"1","t":"call","key":"walk_matter","method":"POST","path":"/v1/matters/abc/walk","query":[["include","1"]],"body":{"choice_id":"c"}}"#
        );
    }

    #[test]
    fn absent_and_null_bodies_stay_distinct() {
        let absent = Frame::decode(br#"{"v":1,"id":"1","t":"end"}"#).unwrap();
        let null = Frame::decode(br#"{"v":1,"id":"1","t":"data","body":null}"#).unwrap();
        assert!(absent.body.is_none());
        assert_eq!(null.body, Some(Value::Null));
    }

    #[test]
    fn version_narrowing_unknown_members_and_duplicate_meta_fail_closed() {
        assert!(Frame::decode(br#"{"v":257,"id":"1","t":"end"}"#).is_err());
        assert!(Frame::decode(
            br#"{"v":1,"id":"1","t":"end","deadline":"5s"}"#
        )
        .is_err());

        let mut duplicate = Frame::end("1");
        duplicate.meta = vec![("x".into(), "1".into()), ("x".into(), "2".into())];
        assert!(duplicate.validate().is_err());

        let replaced = Frame::end("1").with_meta("x", "1").with_meta("x", "2");
        assert_eq!(replaced.meta_get("x"), Some("2"));
        assert_eq!(replaced.meta.len(), 1);
    }

    #[test]
    fn a_corrupt_prefix_is_rejected_before_allocation() {
        let mut buffer = u32::MAX.to_be_bytes().to_vec();
        buffer.extend_from_slice(b"{}");
        assert!(decode_stream(&buffer).is_err());
    }

    #[test]
    fn a_partial_tail_is_retained() {
        let first = Frame::end("1").encode_tcp().unwrap();
        let second = Frame::cancel("2").encode_tcp().unwrap();
        let mut buffer = first;
        buffer.extend_from_slice(&second[..3]);
        let (frames, consumed) = decode_stream(&buffer).unwrap();
        assert_eq!(frames.len(), 1);
        assert_eq!(buffer.len() - consumed, 3);
    }

    #[test]
    fn every_conformance_fixture_round_trips() {
        let raw = include_str!("../../examples/frames/conformance.json");
        let document: Value = serde_json::from_str(raw).expect("fixtures parse");
        let cases = document["cases"].as_array().expect("cases array");
        assert!(!cases.is_empty());
        for case in cases {
            let name = case["name"].as_str().unwrap();
            let encoded = case["encoded"].as_str().unwrap();
            let frame = Frame::decode(encoded.as_bytes())
                .unwrap_or_else(|error| panic!("{name}: decode failed: {error}"));
            assert_eq!(
                String::from_utf8(frame.encode().unwrap()).unwrap(),
                encoded,
                "{name}: canonical bytes"
            );
            assert_eq!(
                hex(&frame.encode_tcp().unwrap()[..LENGTH_PREFIX_BYTES]),
                case["tcp_prefix_hex"].as_str().unwrap(),
                "{name}: length prefix"
            );
        }
    }

    #[test]
    fn correlation_ids_are_monotonic() {
        let mut correlator = Correlator::new("c7-");
        assert_eq!(
            [correlator.take(), correlator.take()],
            ["c7-1", "c7-2"]
        );
    }

    fn hex(bytes: &[u8]) -> String {
        bytes.iter().map(|byte| format!("{byte:02x}")).collect()
    }
}
